lxd/storage/backend_lxd.go

```diff
@@ -844,6 +844,26 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
 			}
 			b.DeleteInstance(inst, op)
 		}()
+
+		// If the negotiated migration method is rsync and the instance's base image is
+		// already on the host then pre-create the instance's volume using the locla image
+		// to try and speed up the rsync of the incoming volume by avoiding the new to
+		// transfer the base image files too.
+		if args.MigrationType.FSType == migration.MigrationFSType_RSYNC {
```

Comment:
Why are we doing this in the backend rather than passing the base image through `args` over to the driver and then have the driver attempt to locate the image and create the volume from that?

This wouldn't even need to hit the database at all and would let us re-use images that have been deleted too (in the case of zfs and ceph where referenced but deleted images are still usable).

Reply:
@stgraber the original code did a DB lookup so I went with that approach in backendLXD so that non optimized storage drivers that cant tell if an image exists (because they dont have their own volumes for images) without looking in the DB can still use this optimization. Have I missed something?

Reply:
@stgraber would it be viable to look directly in the filesystem for the image tarball?

Reply:
Ah, makes sense. Ok, so your approach is probably fine then.

I don't really like the fact that we allow `CreateVolumeFromMigration` on top of an existing volume when refresh isn't passed, but there's nothing technically wrong with it and having the driver directly access the image tarballs would feel even more wrong :)

Reply:
@stgraber there is a check at the top of `CreateInstanceFromMigration` that prevents migrating over an existing volume that we didn't create ourselves during the process. However I could push this down into the driver by in a fairly generic way by doing the image DB lookup still in the backendLXD and if an image was found and Rsync was being used then pass an additional argument to the driver's CreateVolumeFromMigration() function that was a struct called `baseImgFiller` that contained the base image fingerprint as a string and a `fillerFunc` that would then unpack the image using the backendLXD.imageFiller function.

That would achieve the best of both worlds as storage drivers that maintained their own volumes could use the baseImgFiller's fingerprint property to look for any existing image volumes, and any non-volume drivers could use the baseImgFiller's fillerFunc to prepopulate the volume.

What do you think?

Reply:
Sounds good. To make things easier on the driver, we'd probably also want to have the backend run EnsureImage so that if the image doesn't yet exist on a driver which supports optimized storage, then it gets created first.

This means that all drivers that support optimized images can just look for an existing volume and if that doesn't exist, move on and only drivers without optimized images will use the filler.

Reply:
@stgraber implemented this change heree https://github.com/lxc/lxd/pull/6555